docs/metrics-data-model.md

```diff
@@ -0,0 +1,107 @@
+# k6 Time Series
+
+We want to introduce the time series concept to k6 for getting the benefit of efficiently identifying the combination between a metric and the relative tags.
+
+A summary of the benefits it brings across the platform:
+
+* It allows executing faster lookups and comparisons by a single value, the ID (hash) of the series.
+* It reduces the storage by allocating just one time a possible combination (the time series) and referencing it in all other places. 
+* It allows aggregating data by time series.
+* It enables easier detection of high cardinality issues just by counting the entire unique data set of generated time series.
+* It enables time series based outputs to simplify the integration's logic (e.g. Prometheus).
+
+## Data model
+
+#### Metrics and Series
+
+* Metric (name, type)
+* Time Series (metric, tags)
+* Sample (timestamp, value, time series ref)
+* Tag - (aka [k6/1831](https://github.com/grafana/k6/issues/1831))
+
+```go
+// This is the current k6 Metric polished from the Thresholds dependencies.
+type Metric struct {
+	Name     string
+	Type     MetricType
+	Contains ValueType
+}
+
+type Tag struct {
+    Key, Value string
+}
+
+type TimeSeries struct {
+	ID uint64 // hash(metric+tags)
+	MetricName string
+	Tags   []Tag
+}
+
+type Sample struct {
+        TimeSeries *TimeSeries
+        Timestamp uint64
+        Value float64
+}
```

Comment:
>For one thing, we should probably try to make a lot of these things read-only once they have been created.

A common discussion in the Go world. :smile: I'm not convinced, some questions:

* We will create a not consistent situation between Metric and TimeSeries. For example, why is it relevant to keep TimeSeries protected and to not protect the Metric for a name change? If you are suggesting doing the same for Metric then the impact in the source code seems big.
* How should the Get methods work? I expect they have to make **a copy** of the internal structure. It sounds not optimal in terms of efficiency.

>but we should probably also have some index

I agree.

> so the rest of the codebase could efficiently lookup the value of specific tags?

 Out of curiosity, when do you expect to use it? I expect most of the time `TimeSeries.Tags` access should be used for mapping the **entire** slice because a TimeSeries has been already resolved by matching the ID. In that case, the slice iteration would be ok.

I see the index as useful for the `Contains` operation where we need the pair `key+val` as keys otherwise we need to access by key the map but make a string comparison for the value. Right?

Reply:
> For example, why is it relevant to keep TimeSeries protected and to not protect the Metric for a name change? If you are suggesting doing the same for Metric then the impact in the source code seems big.

Yes, I want to fix `Metric` too. I don't think the impact will be that big, it should be a fairly simple substitution. Most extensions don't use the metric name directly, they use a pointer to the `Metric` object. And besides, we make no promises for backwards compatibility of the Go APIs.

> How should the Get methods work? I expect they have to make a copy of the internal structure. It sounds not optimal in terms of efficiency.

I guess depends on the specific thing. For example, for [`metrics.Metric`](https://github.com/grafana/k6/blob/3e21e678fa13f1cdd442877a105a12f0f255c505/metrics/metric.go#L13-L16), I expect getter methods that make copies of its individual properties will be enough: `m.GetName()` instead of just `m.Name`, etc. Besides, [strings in Go are immutable](https://yourbasic.org/golang/gotcha-strings-are-immutable/), so I don't think it even makes an actual copy, just returns a read-only reference to the original string. 

The problem here is that we [always get a pointer to a `Metric`](https://github.com/grafana/k6/blob/3e21e678fa13f1cdd442877a105a12f0f255c505/metrics/registry.go#L52), we can't have copies of the whole struct since the whole point after the recent refactorings is that we only have a single `Metric` object with the same name. So its properties are mutable :disappointed: 

> 
 Out of curiosity, when do you expect to use it? I expect most of the time `TimeSeries.Tags` access should be used for mapping the **entire** slice because a TimeSeries has been already resolved by matching the ID. In that case, the slice iteration would be ok.
> 
> I see the index as useful for the `Contains` operation where we need the pair `key+val` as keys otherwise we need to access by key the map but make a string comparison for the value. Right?

Yeah, mostly for `Contains()`, since that is needed by the `MetricsEngine` for sub-metric sinks in the end-of-test summary and thresholds. Though that can probably be implemented internally :thinking: Still, I can imagine how some output would like the value of a specific tag (e.g. `url`, `method`, etc.) to do some pre- or post-processing only on specific raw metric samples :man_shrugging: In general though, it should be possible to efficiently inspect a `TimeSeries` and know what are its characteristics.

Reply:
It's fine to have a `GetTags()` function, since sometimes you just need a copy of the tags :man_shrugging: As long as we don't have to make these copies in the normal flow of metric samples, it should be fine. But you shouldn't need to make a copy of the whole tag set in order to get the value for a specific tag.

Reply:
>since sometimes you just need a copy of the tags

Maybe I'm not considering something but it's not sometimes, it's every time a `flushMetrics` method is invoked for outputs. We need to map the tags and we don't know any specific tag we just have the Sample and the pointer to the TimeSeries. For example, [this line](https://github.com/grafana/k6/blob/3e21e678fa13f1cdd442877a105a12f0f255c505/output/csv/output.go#L224) would be replaced with the `GetTags()` method, right?